modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -202,21 +215,85 @@ public CompletableFuture<Void> finish(
     ) {
         assert groups != null && !groups.isEmpty();
 
-        HybridTimestamp commitTimestamp = commit ? clock.now() : null;
-
-        TxFinishReplicaRequest req = FACTORY.txFinishReplicaRequest()
-                .txId(txId)
-                .timestampLong(clock.nowLong())
-                .groupId(commitPartition)
-                .groups(groups)
-                .commit(commit)
-                .commitTimestampLong(hybridTimestampToLong(commitTimestamp))
-                .term(term)
-                .build();
-
-        return replicaService.invoke(recipientNode, req)
-                // TODO: IGNITE-20033 TestOnly code, let's consider using Txn state map instead of states.
-                .thenRun(() -> changeState(txId, PENDING, commit ? COMMITED : ABORTED));
+        Function<Void, CompletableFuture<Void>> clo = ignored -> {
```

Comment:
Why do not use `Supplier<CompletableFuture<Void>>`?

Reply:
Because Supplier can not be used in completableFuture.thenCompose